File: borgbutler-core/src/main/java/de/micromata/borgbutler/config/Configuration.java
package de.micromata.borgbutler.config;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.micromata.borgbutler.demo.DemoRepos;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class Configuration {
    private Logger log = LoggerFactory.getLogger(Configuration.class);
    /**
     * Default dir name for restoring archives.
     */
    private static final String RESTORE_DIRNAME = "restore";

    @JsonIgnore
    private File workingDir;
    /**
     * The path of the borg command to use.
     */
    private String borgCommand;

    /**
     * Default is 100 MB (approximately).
     */
    private int maxArchiveContentCacheCapacityMb = 100;

    private boolean showDemoRepos = true;

    /**
     * Default is restore inside BorgButler's home dir (~/.borgbutler/restore).
     */
    @JsonProperty("restoreDir")
    private String restoreDirPath;
    @JsonIgnore
    private File restoreHomeDir;

    @JsonProperty
    private List<BorgRepoConfig> repoConfigs = new ArrayList<>();

    public void add(BorgRepoConfig repoConfig) {
        synchronized (repoConfigs) {
            repoConfigs.add(repoConfig);
        }
    }

    public boolean remove(String idOrName) {
        if (idOrName == null) {
            return false;
        }
        synchronized (repoConfigs) {
            for (BorgRepoConfig repoConfig : getAllRepoConfigs()) {
                if (StringUtils.equals(idOrName, repoConfig.getRepo()) || StringUtils.equals(idOrName, repoConfig.getId())) {
                    repoConfigs.remove(repoConfig);
                    return true;
                }
            }
        }
        return false;
    }

    public BorgRepoConfig getRepoConfig(String idOrName) {
        if (idOrName == null) {
            return null;
        }
        for (BorgRepoConfig repoConfig : getAllRepoConfigs()) {
            if (StringUtils.equals(idOrName, repoConfig.getRepo()) || StringUtils.equals(idOrName, repoConfig.getId())) {
                return repoConfig;
            }
        }
        return null;
    }

    public File getRestoreHomeDir() {
        if (restoreHomeDir == null) {
            if (StringUtils.isNotBlank(restoreDirPath)) {
                restoreHomeDir = new File(restoreDirPath);
            } else {
                restoreHomeDir = new File(workingDir, RESTORE_DIRNAME);
            }
            if (!restoreHomeDir.exists()) {
                log.info("Creating dir '" + restoreHomeDir.getAbsolutePath() + "' for restoring backup files and directories.");
            }
        }
        return restoreHomeDir;
    }

    public void copyFrom(Configuration other) {
        this.borgCommand = other.borgCommand;
        this.maxArchiveContentCacheCapacityMb = other.maxArchiveContentCacheCapacityMb;
        this.showDemoRepos = other.showDemoRepos;
    }

    @JsonIgnore
    public List<BorgRepoConfig> getAllRepoConfigs() {
        return DemoRepos.getAllRepos(repoConfigs);
    }

    List<BorgRepoConfig> getRepoConfigs() {
        return repoConfigs;
    }

    public String getBorgCommand() {
        return this.borgCommand;
    }

    public int getMaxArchiveContentCacheCapacityMb() {
        return this.maxArchiveContentCacheCapacityMb;
    }

    public boolean isShowDemoRepos() {
        return this.showDemoRepos;
    }

    public String getRestoreDirPath() {
        return this.restoreDirPath;
    }

    Configuration setWorkingDir(File workingDir) {
        this.workingDir = workingDir;
        return this;
    }

    public Configuration setBorgCommand(String borgCommand) {
        this.borgCommand = borgCommand;
        return this;
    }

    public Configuration setShowDemoRepos(boolean showDemoRepos) {
        this.showDemoRepos = showDemoRepos;
        return this;
    }
}

File: borgbutler-core/src/main/kotlin/de/micromata/borgbutler/config/Configuration.kt
package de.micromata.borgbutler.config

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import de.micromata.borgbutler.config.BorgRepoConfig
import de.micromata.borgbutler.demo.DemoRepos
import mu.KotlinLogging
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import java.io.File

private val log = KotlinLogging.logger {}

/**
 * Representation of ~/.borgbutler/borgbutler-config.json.
 */
open class Configuration {

    @JsonIgnore
    private var workingDir: File? = null
    /**
     * Optional borg command to use.
     */
    /**
     * The path of the borg command to use (optional).
     */
    var borgCommand: String? = null
        private set

    /**
     * The borg version to install from github (optional).
     */
    var borgVersion: String? = null

    /**
     * Default is 100 MB (approximately).
     */
    var maxArchiveContentCacheCapacityMb = 100
        private set
    var isShowDemoRepos = true
        private set

    /**
     * Default is restore inside BorgButler's home dir (~/.borgbutler/restore).
     */
    @JsonProperty("restoreDir")
    val restoreDirPath: String? = null

    @JsonIgnore
    private var restoreHomeDir: File? = null

    @JsonProperty
    private val repoConfigs = mutableListOf<BorgRepoConfig>()

    fun add(repoConfig: BorgRepoConfig) {
        synchronized(repoConfigs) { repoConfigs.add(repoConfig) }
    }

    fun remove(idOrName: String?): Boolean {
        if (idOrName == null) {
            return false
        }
        synchronized(repoConfigs) {
            for (repoConfig in allRepoConfigs) {
                if (StringUtils.equals(idOrName, repoConfig.repo) || StringUtils.equals(idOrName, repoConfig.id)) {
                    repoConfigs.remove(repoConfig)
                    return true
                }
            }
        }
        return false
    }

    fun getRepoConfig(idOrName: String?): BorgRepoConfig? {
        if (idOrName == null) {
            return null
        }
        for (repoConfig in allRepoConfigs) {
            if (StringUtils.equals(idOrName, repoConfig.repo) || StringUtils.equals(idOrName, repoConfig.id)) {
                return repoConfig
            }
        }
        return null
    }

    fun getRestoreHomeDir(): File {
        if (restoreHomeDir == null) {
            restoreDirPath?.let {
                restoreHomeDir = File(it)
            } ?: run {
                restoreHomeDir = File(workingDir, RESTORE_DIRNAME)
            }
            if (!restoreHomeDir!!.exists()) {
                log.info("Creating dir '" + restoreHomeDir?.absolutePath + "' for restoring backup files and directories.")
            }
        }
        return restoreHomeDir!!
    }

    fun copyFrom(other: Configuration) {
        borgCommand = other.borgCommand
        maxArchiveContentCacheCapacityMb = other.maxArchiveContentCacheCapacityMb
        isShowDemoRepos = other.isShowDemoRepos
    }

    @get:JsonIgnore
    val allRepoConfigs: List<BorgRepoConfig>
        get() = DemoRepos.getAllRepos(repoConfigs)

    fun getRepoConfigs(): List<BorgRepoConfig> {
        return repoConfigs
    }

    fun setWorkingDir(workingDir: File?): Configuration {
        this.workingDir = workingDir
        return this
    }

    fun setBorgCommand(borgCommand: String?): Configuration {
        this.borgCommand = borgCommand
        return this
    }

    fun setShowDemoRepos(showDemoRepos: Boolean): Configuration {
        isShowDemoRepos = showDemoRepos
        return this
    }

    companion object {
        /**
         * Default dir name for restoring archives.
         */
        private const val RESTORE_DIRNAME = "restore"
    }
}
